python_modules/dagster/dagster_tests/declarative_automation_tests/scenario_utils/asset_daemon_scenario.py

```diff
@@ -199,22 +200,42 @@ def _evaluate_tick_daemon(
                 # start sensor if it hasn't started already
                 self.instance.start_sensor(sensor)
 
-            amp_tick_futures = {}
-
-            list(
-                AssetDaemon(  # noqa: SLF001
-                    settings=self.instance.get_auto_materialize_settings(),
-                    pre_sensor_interval_seconds=42,
-                )._run_iteration_impl(
-                    workspace_context,
-                    threadpool_executor=self.threadpool_executor,
-                    amp_tick_futures=amp_tick_futures,
-                    debug_crash_flags={},
-                    submit_threadpool_executor=None,
+            def _stop_sensor():
+                if sensor:
+                    self.instance.stop_sensor(
+                        sensor.get_remote_origin_id(), sensor.selector_id, sensor
+                    )
+
+            def _run_daemon():
+                amp_tick_futures = {}
+
+                list(
+                    AssetDaemon(  # noqa: SLF001
+                        settings=self.instance.get_auto_materialize_settings(),
+                        pre_sensor_interval_seconds=42,
+                    )._run_iteration_impl(
+                        workspace_context,
+                        threadpool_executor=self.threadpool_executor,
+                        amp_tick_futures=amp_tick_futures,
+                        debug_crash_flags={},
+                        submit_threadpool_executor=None,
+                        poll_interval=0,  # check the status of the sensor after every run submission
+                    )
                 )
-            )
 
-            wait_for_futures(amp_tick_futures)
+                wait_for_futures(amp_tick_futures)
+
+            if stop_mid_iteration:
+                test_futures = {}
```

Comment:
this approach of stopping the sensor in another thread makes me a bit nervous since it does rely a bit on the timing of the stopping thread to actually stop the sensor at the right time (after the first time we check the status of the sensor right at the beginning of the evaluation, but before we submit any runs). It's the best option i could cook up though, but i'd be open to other ideas!

Reply:
one gross idea would be to attach an automation condition to the asset that sleeps until the sensor has stopped..